solr/core/src/test/org/apache/solr/cloud/overseer/ZkStateReaderTest.java

```diff
@@ -609,4 +612,60 @@ public void testWatchRaceCondition() throws Exception {
       ExecutorUtil.awaitTermination(executorService);
     }
   }
+
+  /**
+   * Ensure that `fetchCollectionState` would not throw exception when the state.json is deleted in between the
+   * state.json read and PRS entries read
+   * @throws Exception
+   */
+  public void testDeletePrsCollection() throws Exception {
+    ZkStateWriter writer = fixture.writer;
+    ZkStateReader reader = fixture.reader;
+
+    fixture.zkClient.makePath(ZkStateReader.COLLECTIONS_ZKNODE + "/c1", true);
+
+    ClusterState clusterState = reader.getClusterState();
+    // create new collection
+    DocCollection state =
+        new DocCollection(
+            "c1",
+            new HashMap<>(),
+            Collections.singletonMap(DocCollection.PER_REPLICA_STATE, true),
+            DocRouter.DEFAULT,
+            0,
+            ZkStateReader.CLUSTER_STATE + "/c1/state.json");
+    ZkWriteCommand wc = new ZkWriteCommand("c1", state);
+    writer.enqueueUpdate(clusterState, Collections.singletonList(wc), null);
+    writer.writePendingUpdates();
+
+    TimeOut timeOut = new TimeOut(5000, TimeUnit.MILLISECONDS, TimeSource.NANO_TIME);
+    timeOut.waitFor(
+        "Timeout on waiting for c1 to show up in cluster state",
+        () -> reader.getClusterState().getCollectionOrNull("c1") != null);
+
+    String collectionPath = ZkStateReader.getCollectionPath("c1");
+    ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor(new SolrNamedThreadFactory("zk-state-reader-test-scheduler"));
+    executor.schedule(() -> {
+      try {
+        reader.getZkClient().clean(collectionPath); //call this directly, very tricky to control execution order with writer.enqueueUpdate
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } catch (KeeperException e) {
+        throw new RuntimeException(e);
+      }
+    },100, TimeUnit.MILLISECONDS);
```

Comment:
iiuc the purpose of the delay here, is 100ms enough to guarantee that the collection deletion will happen in between when the subsequent request fetches `state.json` and when it fetches the PRS state? Put another way, is it possible that the check below might just say "yup, the collection's been deleted" and not actually be exercising the "state.json ok, PRS missing" case that we're interested in?

Reply:
Addressed with the new Breakpoint approach 🎉